testcases/kernel/syscalls/adjtimex/adjtimex02.c

```diff
@@ -10,115 +10,188 @@
+	},
+	{
+	.modes = ADJ_TICK,
+	.highlimit = 1100000,
+	.delta = 1,
+	.exp_err = EINVAL,
+	},
+	{
+	.modes= SET_MODE,
+	.exp_err = EPERM,
+	.droproot = 1,
+	},
+	{
+	.modes = ADJ_OFFSET,
+	.highlimit = 512000L,
+	.delta = 1,
+	.exp_err = EINVAL,
+	},
+	{
+	.modes = ADJ_OFFSET,
+	.lowlimit = -512000L,
+	.delta = -1,
+	.exp_err = EINVAL,
+	},
+};
+
+static struct test_variants
+{
+	int (*adjtimex) (void *timex);
+	enum tst_timex_type type;
+	char *desc;
+} variants[] = {
+#if (__NR_adjtimex != __LTP__NR_INVALID_SYSCALL)
+	{
+	.adjtimex = sys_adjtimex,.type = TST_KERN_OLD_TIMEX,.desc = "syscall with old kernel spec"},
+#endif
+};
+
+static void verify_adjtimex(unsigned int i)
 {
 	struct timex *bufp;
-	int expected_errno = 0;
-
-	/*
-	 * since Linux 2.6.26, if buf.offset value is outside
-	 * the acceptable range, it is simply normalized instead
-	 * of letting the syscall fail. so just skip this test
-	 * case.
-	 */
-	if (nr > 3 && (tst_kvercmp(2, 6, 25) > 0)) {
-		tst_res(TCONF, "this kernel normalizes buf."
-				"offset value if it is outside"
-				" the acceptable range.");
-		return;
-	}
+	struct test_variants *tv = &variants[tst_variant];
 
 	*buf = *tim_save;
 	buf->modes = SET_MODE;
-	bufp = buf;
-	switch (nr) {
-	case 0:
-		bufp = (struct timex *)-1;
-		expected_errno = EFAULT;
-		break;
-	case 1:
-		buf->tick = 900000 / hz - 1;
-		expected_errno = EINVAL;
-		break;
-	case 2:
-		buf->tick = 1100000 / hz + 1;
-		expected_errno = EINVAL;
-		break;
-	case 3:
+	bufp =buf;
+
+	if(tc[i].droproot) {
 		/* Switch to nobody user for correct error code collection */
 		ltpuser = SAFE_GETPWNAM("nobody");
 		SAFE_SETEUID(ltpuser->pw_uid);
-		expected_errno = EPERM;
-		break;
-	case 4:
-		buf->offset = 512000L + 1;
-		expected_errno = EINVAL;
-		break;
-	case 5:
-		buf->offset = (-1) * (512000L) - 1;
-		expected_errno = EINVAL;
-		break;
-	default:
-		tst_brk(TFAIL, "Invalid test case %u ", nr);
 	}
 
-	TEST(adjtimex(bufp));
-	if ((TST_RET == -1) && (TST_ERR == expected_errno)) {
-		tst_res(TPASS | TTERRNO,
-				"adjtimex() error %u ", expected_errno);
-	} else {
-		tst_res(TFAIL | TTERRNO,
-				"Test Failed, adjtimex() returned %ld",
-				TST_RET);
+	if(tc[i].exp_err == EINVAL) {
+		if(tc[i].modes == ADJ_TICK) {
+			if(tc[i].lowlimit)
+				buf->tick = tc[i].lowlimit - tc[i].delta;
+
+			if(tc[i].highlimit)
+				buf->tick = tc[i].highlimit + tc[i].delta;
+		}
+		if(tc[i].modes == ADJ_OFFSET) {
+			if(tc[i].lowlimit) {
+				tst_res(TCONF, "this kernel normalizes buf."
+				"offset value if it is outside"
+				" the acceptable range.");
+				return;
+			}
+			if(tc[i].highlimit) {
+				tst_res(TCONF, "this kernel normalizes buf."
+				"offset value if it is outside"
+				" the acceptable range.");
+				return;
+			}
+		}
+	}
+
+	if(tc[i].exp_err == EFAULT ) {
+			tst_res(TINFO, "EFAULT is tested only for raw syscall");
+			bufp = (struct timex *) -1;
+			TEST (tv->adjtimex (bufp));
+		} else {
+			TEST(tv->adjtimex(bufp));
+			TEST(adjtimex(bufp));
 	}
```

Comment:
Looks like moving the test data into the struct testcase structure made the code much worse. Why don't we just add the variants and keep the code mostly as it was? I guess that all that will be needed would be calling the tv->adjtimex() instead.

Reply:
Corrections made.